core-plugins/src/main/java/co/cask/hydrator/plugin/batch/action/SSHAction.java

```diff
@@ -61,12 +62,19 @@ public SSHAction(SSHActionConfig config) {
 
   @Override
   public void run(final ActionContext context) throws Exception {
+    // now that macros have been substituted, try validation again
+    config.validate();
+
     Connection connection = new Connection(config.host);
     try {
       connection.connect();
 
-      String password = Strings.isNullOrEmpty(config.password) ? null : config.password;
-      if (!connection.authenticateWithPublicKey(config.user, new File(config.privateKeyFile), password)) {
+      // get private key from secure store using user-specified key
+      char[] privateKey =
+        new String(context.getSecureData(context.getNamespace(), config.privateKeyFileLookupKey).get()).toCharArray();
```

Comment:
Should specify charset while converting from ``byte[]`` to ``String``. Use ``Bytes.toString(...)``.